From b82e5e2ae143a0ce48ec800fbc9dfcf0a7c95311 Mon Sep 17 00:00:00 2001
From: Linn Crosetto <linn@hpe.com>
Date: Thu, 10 Mar 2016 18:25:03 -0700
Subject: [PATCH] acpi: Disable ACPI table override when UEFI Secure Boot is
 enabled

Patch-mainline: Not yet, Need to send with other secure boot lock down patches
References: bsc#970604

From the kernel documentation (initrd_table_override.txt):

  If the ACPI_INITRD_TABLE_OVERRIDE compile option is true, it is possible
  to override nearly any ACPI table provided by the BIOS with an
  instrumented, modified one.

Do not allow ACPI tables to be overridden if UEFI Secure Boot is enabled.

Signed-off-by: Linn Crosetto <linn@hpe.com>
Acked-by: Lee, Chun-Yi <jlee@suse.com> 
---
 arch/x86/kernel/setup.c |   16 ++++++++--------
 drivers/acpi/osl.c      |    6 ++++++
 2 files changed, 14 insertions(+), 8 deletions(-)

--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -1121,6 +1121,14 @@ void __init setup_arch(char **cmdline_p)
 	/* Allocate bigger log buffer */
 	setup_log_buf(1);
 
+#ifdef CONFIG_EFI_SECURE_BOOT_SIG_ENFORCE
+	if (boot_params.secure_boot) {
+		set_bit(EFI_SECURE_BOOT, &efi.flags);
+		enforce_signed_modules();
+		pr_info("Secure boot enabled\n");
+	}
+#endif
+
 	reserve_initrd();
 
 #if defined(CONFIG_ACPI) && defined(CONFIG_BLK_DEV_INITRD)
@@ -1131,14 +1139,6 @@ void __init setup_arch(char **cmdline_p)
 
 	io_delay_init();
 
-#ifdef CONFIG_EFI_SECURE_BOOT_SIG_ENFORCE
-	if (boot_params.secure_boot) {
-		set_bit(EFI_SECURE_BOOT, &efi.flags);
-		enforce_signed_modules();
-		pr_info("Secure boot enabled\n");
-	}
-#endif
-
 	/*
 	 * Parse the ACPI tables for possible boot-time SMP configuration.
 	 */
--- a/drivers/acpi/osl.c
+++ b/drivers/acpi/osl.c
@@ -629,6 +629,12 @@ void __init acpi_initrd_override(void *d
 	if (table_nr == 0)
 		return;
 
+	if (secure_modules()) {
+		pr_notice(PREFIX
+			"security enabled, ignoring table override\n");
+		return;
+	}
+
 	acpi_tables_addr =
 		memblock_find_in_range(0, max_low_pfn_mapped << PAGE_SHIFT,
 				       all_tables_size, PAGE_SIZE);
